### ISSUE TITLE
Make withAuthenticator Composable

### ISSUE THREAD

**Issue by pigmanbear** [aws-amplify/amplify-js#1608]
**Is your feature request related to a problem? Please describe.**
In order to compose withAuthenticator a helper function must be created e.g.: 

```
const withAmpAuth = x => withAuthenticator(x, ...)
```
in order to 
```
export default compose(withThis, withThat, withAmpAuth)(Component)
```
And also pass in an object instead of args
**Describe the solution you'd like**
Instead of :
```
export function withAuthenticator(Comp, includeGreetings = false, authenticatorComponents = [], federated = null, theme = null){ ... }
```
Perhaps: 
 ```
export const withAuthenticator = ({includeGreetings: false, authenticatorComponents: [], federated: null, theme: null}) => Component => { ... }
```
or I guess if we are using 'function':
```
export function withAuthenticator({includeGreetings: false, authenticatorComponents: [], federated: null, theme: null}) { return  function (Component) {return ... } }
```
then using withAuthenticator can be part compose, and arity is just two, with object and Component

```
export default  compose(withThis, withThat, withAmpAuth(object))(Component)
```

**Comment by undefobj** [comment]
@pigmanbear can you give an example of why you would need to wrap more than one component for an authentication routine? It's almost always the top level route of your UX. To take your example:

`export default compose(withThis, withThat, withAmpAuth)(Component)
`
What would `withThis` or `withThat` do in an application?

**Comment by pigmanbear** [comment]
@undefobj Certainly, I can share some scenarios.

Subsequently, I have taken a deep dive into the project, viewing docs and source. I have implemented custom components for the sign in flow, etc. using the Authenticator component as opposed the HOC to start. Now, I have started working on rolling some of my own, favoring composition over inheritance and components doing a little less work.

I will also preface and say I avoid using stateful components and prefer the stateless functional variety then use recompose, redux, reselect, and more recently react-streams, for managing state, lifecyle updates, etc. so many of components have a compose( connect(), withHandlers, withLifecylce, withRouter, branch, withContext, defaultProps, withOther, withThat, withThis)(Component) flavor, or a stream with many operators in pipe and passed to the component. Well, maybe not quite that bad, but you probably get the point.  

With all that said, withAuthenticator, the naming is a little confusing with respect to many other packages in the community. In most React packages a 'withSomething' is usually included for Consumer/Provider with respect to the new context API, or access outside of a component tree, and doesn't necessarily bring as much with it. 

As for the specific use case was a protecting routes, and even individual components based on the logged in user.  

Or even in a menu bar to hide/display links.  Or even just getting the authState/authData passed and using custom logic. 

I also have some pages that are 'mixed use' where an authenticated user could see more data or more contextual information on again an otherwise unauthorized route. 

First I went with sometthing like:

```js
const AuthenticatedRoute = ({ component: Component, ...rest }) => (
  <Route
    {...rest}
    render={props => (
      <Authenticator theme={MyTheme} >
        <Component {...props} />
      </Authenticator>
    )}
  />
```
Then build Routes with:
```js
            <Route path="/route1" component={Component1} />
            <AuthenticatedRoute path="/route2" component={Component2} />
```
But, the show/hide wasn't working well, and re-renders seemed excessive, so I grabbed withAuthenticator without reading the documentation too much and seeing it does most the same thing, which I should have noticed before posting this, but didn't really take a look at of the docs or even look at the component all the way through. 

The mixed use case, using auth to show/hide components outside of an authenticated component tree, and not necessarily wanting Authentication at the root level but having access is the reason I was looking to use withAuthenticator. 

So, I think withAuthenticator is really just named 'unconventionally' considering it is doing much more than providing auth context. 

Still, loving using Amplify, and love the project, and just adding this comment to at least let you know the use cases and also suggest maybe a withAuthenticator that doesn't do so much and more inline with the other 'withs' in the wild. 

Again, awesome project and look forward to seeing it continue to improve and evolve.

**Comment by pigmanbear** [comment]
After customizing all of the Components in amplify-react by inheriting and using showComponent, etc, and then adding why-did-you-update to take a look at the renders (which can be seen also by setting log level to debug, among others):
<img width="633" alt="screenshot_2018-09-14 09 03 55_3armxv" src="https://user-images.githubusercontent.com/32139940/45574570-b1c02980-b82d-11e8-825b-710d104c8a75.png">
, considering these components are not lazy loaded, and loaded even when using dynamic imports and greatly increasing bundle size and making code splitting a pain, I am working using the createContext and Provider/Consumer/hoc pattern to implement @aws-amplify/auth. 

And I really did appreciate the idea behind the Authenticator/children/top-level-wrapper for not worrying about redirects, for signin and signout URLs, forgetting, resetting, etc. So, I experimented with using branch from recompose: 

```js
export default compose(
   withAuth,
   ....,
  ConfirmSignIn(props => props.authState === 'confirmSignIn'),
  RequireNewPassword(props => props.authState === 'requireNewPassword'),
  SignIn(props => props.authState === 'signin')
)(SomeComponent)
```
where SignIn, etc. is defined:
```js
const SignIn = authState => branch(authState, renderComponent(MaterialSignIn)) 
```
MaterialSignIn is the dumb functional component and receiving props from the withAuth context Consumer. The Provider is similar to the AuthPiece Component, with a few pieces from the other lib Components (though implemented with react-streams).

And now, I really don't need to necessarily have authenticated routes. In a single component, if the authState is not equal to a defined enum, then the default authState can be 'unauth', or strictly authenticated routes can also make use of branch, and add the 'right' component to render an access error or renderNothing, as a final barrier (though without a token, nothing worthwhile would be loaded anyhow).  Or, the mixed-use component children can also check props for authState to determine if a signed-in user contextual information will be rendered.  Still testing and working on it, but looks like a viable solution so far, and thinking about it along the same lines as was implemented in amplify-react (no redirects, or urls for auth flow), but without the inheritance and eager loading

Now, if I just want to use withAuth, I can just wrap a component, which I was thinking would come in handy in a few scenarios, like the Remember Me functionality I was able to get working using the SDK, though I was toying around with Hub and think that might be the best solution for that one: 

```js
alex.onHubCapsule = capsule => {
  of(capsule)
    .pipe()
    .subscribe(async x => {
      window.localStorage.setItem('alex', JSON.stringify(x))
      let cognitoUser
      if ((x.payload.event = 'signIn')) {
        try {
          cognitoUser = await Auth.currentUserPoolUser()
          const keyPrefix = `CognitoIdentityServiceProvider.${
            cognitoUser.pool.clientId
          }.${cognitoUser.attributes.email}`
          const deviceKey = localStorage.getItem(`${keyPrefix}.deviceKey`)
          const accessToken = localStorage.getItem(`${keyPrefix}.accessToken`)
          if (deviceKey && accessToken) {
            await Auth.currentCredentials()
              .then(credentials => {
                const cisp = new CognitoIdentityServiceProvider({
                  apiVersion: '2016-04-18',
                  region: 'us-east-2',
                  credentials: Auth.essentialCredentials(credentials),
                })
                cisp
                  .updateDeviceStatus({
                    AccessToken: accessToken,
                    DeviceKey: deviceKey,
                    DeviceRememberedStatus: 'remembered',
                  })
                  .promise()
                  .then(x => logger.debug(x))
                  .catch(y => logger.debug(y))
              })
              .catch(y => logger.debug(y))
          }
        } catch (e) {}
      }
    })
}
```
With Hub, and if implementing redux, could dispatch an action and add the user/userState to the store, and a connected component could then show or hide based on the store vs the createContext Provider/Consumer . . . so many options! 

Still early on in putting this together for the app I am building, which has a relatively aggressive timeline for production. So, it's great there are some defaults to fall back on :)

I apologize for the rambling, thought it worthwhile to add some more information/context.
